### PR TITLE
Move numeric consts to associated consts step1

### PR DESCRIPTION
**Pull request by faern**
A subset of #67913. Implements the first step of RFC https://github.com/rust-lang/rfcs/pull/2700

This PR adds the new constants as unstable constants and defines the old ones in terms of the new ones. Then fix a tiny bit of code that started having naming collisions because of the new assoc consts.

Removed a test that did not seem relevant any longer. Since doing just `u8::MIN` should now indeed be valid.

### CODE REVIEW

**Review by dtolnay — COMMENTED** [review]
Seeing new name collisions is somewhat concerning. Could you dig into the situation in which it's happening and how common that might be in downstream code?

**Comment by faern** [comment]
@dtolnay Adding an associated constant directly on a type which implements a trait that has an associated constant with the same name (in scope) silently makes `TheType::THE_CONST` refer to the newly added constant directly on the type, yes. Here is a playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=21f22f4f5a4e5df7114274e7ba0371ac

So for this PR/RFC that means anyone implementing traits on the primitive numeric types that have constants with the same names as the ones we add will now get our new values unless they properly wrote `<f32 as TheTrait>::EPSILON` instead of just `f32::EPSILON`.

I have not done any research into how common that might be downstream. I have no idea how I would do that except for a crater run? I hope/assume it would not be a common problem. Because even if some people have defined traits with these constants on the primitive types they are very likely the same type and the same value, it would not make sense to add these constants with different values really. And if they do have different types (like the errors I had to fix in this PR) then they will at least get a compilation error and not silently new values.

The exact same thing happens if you add a function to a type that has a function with the same name defined in a trait. Neither of these can be considered breaking(?!) because that would mean you could never add constants or functions to any type, because you risk overriding similar items on traits implemented by that type.

**Comment by faern** [comment]
Could someone help me fix/understand the broken `show_const_contents` test failure? I don't know how to run it locally. And I'm not fully sure what it tests even. It looks like it tests if the html version of generated docs contain the constants. However, when I look at documentation I never see the actual values, I just see `pub const EPSILON: f32`

**Comment by faern** [comment]
Fixed. `./x.py test --stage 1 src/test/rustdoc` and looking inside `build/<triple>/test/rustdoc/show-const-contents/` did the job. Just writing that down here in case someone else stumbles upon it. The build system is not really intuitive :)

File: src/libcore/num/mod.rs [hunk 1]
```diff
@@ -245,42 +245,54 @@ macro_rules! int_impl {
...
-            concat!("Returns the smallest value that can be represented by this integer type.
+            concat!("The smallest value that can be represented by this integer type.
 
 # Examples
 
 Basic usage:
 
 ```
-", $Feature, "assert_eq!(", stringify!($SelfT), "::min_value(), ", stringify!($Min), ");",
+", $Feature, "assert_eq!(", stringify!($SelfT), "::MIN, ", stringify!($Min), ");",
```
Comment (LukasKalbertodt) [inline src/libcore/num/mod.rs]: To fix the latest CI failure, you have to add `#![feature(assoc_int_consts)]` somewhere here. The doc comments also need to specify which unstable features they use.

File: src/libcore/num/mod.rs [hunk 2]
```diff
@@ -245,42 +245,54 @@ macro_rules! int_impl {
...
-            concat!("Returns the largest value that can be represented by this integer type.
+            concat!("The largest value that can be represented by this integer type.
 
 # Examples
 
 Basic usage:
 
 ```
-", $Feature, "assert_eq!(", stringify!($SelfT), "::max_value(), ", stringify!($Max), ");",
+", $Feature, "assert_eq!(", stringify!($SelfT), "::MAX, ", stringify!($Max), ");",
```
Comment (LukasKalbertodt) [inline src/libcore/num/mod.rs]: Same here

**Review by LukasKalbertodt — COMMENTED** [review]
The potential breakage is indeed not very nice. I'm pretty sure this is covered by [RFC 1105 API Evolution](https://github.com/nox/rust-rfcs/blob/master/text/1105-api-evolution.md) and thus allowed. We should still take care, run crater and decide in the team. We'll do that once PR is green.

File: src/libcore/num/mod.rs (hunk 1 above)
Comment (faern) [reply src/libcore/num/mod.rs]: Oh. Is that where those came from. I failed to realize the documentation examples does not compile as the libcore/libstd crates and thus need their own. Thanks!

**Comment by bstrie** [comment]
Looks like this PR is passing, @LukasKalbertodt is now a good time to begin a crater run?

**Comment by faern** [comment]
Seems to pass. So do we want a crater run on this before merging?

**Comment by Mark-Simulacrum** [comment]
@craterbot check

**Comment by bstrie** [comment]
@Mark-Simulacrum I think crater has become stuck, its percentage of completion hasn't updated all day.

**Comment by Mark-Simulacrum** [comment]
Yeah, it does that unfortunately. Cc @pietroalbini

**Comment by faern** [comment]
>  0 regressed and 1 fixed (88601 total)

Sounds pretty optimal to me :) Merge?

**Comment by LukasKalbertodt** [comment]
Yep, that sounds optimal indeed. 

Adding inherent items [is considered a *minor* change by RFC 1105](https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md#minor-change-adding-any-inherent-items). The particular items here are unlikely to be added by user code and even more unlikely to be added with a different meaning. I was thinking this might clash with `num-traits`, but that crate uses associated functions and not consts. We also have no indication that this change would break any Rust code in the wild. Finally, this just follows the plan laid out in the RFC, which has been approved by the team not too long ago.

With all that, I think the teams approval is not necessary in this case.

@bors r+

**Comment by LukasKalbertodt** [comment]
Today I learned: do not edit your comment with the `@bors` approval. Bors always listens. Sorry lovely bot for confusing you!

**Pull request merged**